### PR TITLE
Bucket notifications Get/Set/Listen/RemoveAll operations

### PR DESCRIPTION
**Pull request by BigUstad**
With GetBucketNotificationArgs, SetBucketNotificationArgs, ListenBucketNotificationArgs, RemoveAllBucketNotificationsArgs

### CODE REVIEW

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 1]
```diff
@@ -198,4 +201,147 @@ public override RestRequest BuildRequest(RestRequest request)
...
+                            Console.WriteLine("== ListenBucketNotificationsAsync read line ==");
+                            Console.WriteLine(line);
+                            Console.WriteLine("==============================================");
+                        }
+                        if (line == null)
+                        {
+                            break;
+                        }
+                        string trimmed = line.Trim();
+                        if (trimmed.Length > 2)
```
Comment (krisis) [inline Minio/DataModel/BucketOperationsArgs.cs:312]: what does this `trimmed.Length` check filter out?

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 2]
```diff
@@ -298,6 +298,77 @@ public IObservable<VersionItem> ListObjectVersionsAsync(ListObjectsArgs args, Ca
...
+                    int i = 0;
+                    using (var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ct))
+                    {
+                        while (isRunning)
+                        {
+                            args = args.WithNotificationObserver(obs)
+                                       .WithEnableTrace(this.trace);
+                            RestRequest request = await this.CreateRequest(args).ConfigureAwait(false);
+                            await this.ExecuteTaskAsync(this.NoErrorHandlers, request, cancellationToken).ConfigureAwait(false);
+                            Console.WriteLine(i++);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:364]: Yes Indeed. Removed.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 3]
```diff
@@ -298,6 +298,77 @@ public IObservable<VersionItem> ListObjectVersionsAsync(ListObjectsArgs args, Ca
...
+        {
+            RestRequest request = await this.CreateRequest(args).ConfigureAwait(false);
+            await this.ExecuteTaskAsync(this.NoErrorHandlers, request, cancellationToken).ConfigureAwait(false);
+        }
+
+
+        /// <summary>
+        /// Subscribes to bucket change notifications (a Minio-only extension)
+        /// </summary>
+	    /// <param name="args">ListenBucketNotificationsArgs Arguments Object with information like Bucket name, listen events, prefix filter keys, suffix fileter keys</param>
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:345]: Corrected spelling.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 1 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:312]: To filter out empty stream lines with just newline whitespaces.

**Review by krisis — APPROVED** [review]

**Review by Praveenrajmani — APPROVED** [review]
There are some warnings reported by github actions, Are we planning to fix this separately in future? @BigUstad 

LGTM otherwise

**Comment by BigUstad** [comment]
> There are some warnings reported by github actions, Are we planning to fix this separately in future? @BigUstad
> 
> LGTM otherwise

Yes. We're trying to get all the functionality merged. Most of the warnings were either present before we started with the Builder code or were a result of the functions rendered obsolete. All obsoleted functions are removed in Example tests & Functional tests. And the obsoleted functions also in-turn call the updated functions ensuring backward compatibility.
Thanks :-)

**Pull request merged**